bin/main.ml

```diff
@@ -1121,23 +1120,54 @@ let install_uninstall ~what =
        in
        Fiber.parallel_iter install_files_by_context
          ~f:(fun (context, install_files) ->
-           let install_files_set = Path.Set.of_list install_files in
            get_prefix context ~from_command_line:prefix_from_command_line
            >>= fun prefix ->
            get_libdir context ~libdir_from_command_line
-           >>= fun libdir ->
-           Fiber.parallel_iter install_files ~f:(fun path ->
-             let purpose = Process.Build_job install_files_set in
-             Process.run ~purpose ~env:setup.env Strict opam_installer
-               ([ sprintf "-%c" what.[0]
-                ; Path.to_string path
-                ; "--prefix"
-                ; Path.to_string prefix
-                ] @
-                match libdir with
-                | None -> []
-                | Some p -> [ "--libdir"; Path.to_string p ]
-               ))))
+           >>| fun libdir ->
+           List.iter install_files ~f:(fun (package, path) ->
+             let entries = Install.load_install_file path in
+             let paths =
+               Install.Section.Paths.make
+                 ~package
+                 ~destdir:prefix
+                 ?libdir
+                 ()
+             in
+             let files_deleted_in = ref Path.Set.empty in
+             List.iter entries ~f:(fun { Install.Entry. src; dst; section } ->
+               let src = src in
+               let dst = Option.value dst ~default:(Path.basename src) in
+               let dst =
+                 Path.relative (Install.Section.Paths.get paths section) dst
+               in
+               let dir = Path.parent_exn dst in
+               if what = "install" then begin
+                 Printf.eprintf "Installing %s as %s\n%!"
+                   (Path.to_string src)
+                   (Path.to_string_maybe_quoted dst);
+                 Path.mkdir_p dir;
+                 Io.copy_file () ~src ~dst
+                   ~chmod:(match section with
+                     | Libexec -> (fun x -> x lor 0o111)
+                     | _ -> (fun x -> x))
```

Comment:
[`opam-installer` will clear the executable bit](https://github.com/ocaml/opam/blob/10fcb3f1227db53a256e821cfe9d5f96c66fa368/src/tools/opam_installer.ml#L184) in the other cases, so it should be something like `x land 0o666`. (I'd recommend extracting `set_executable_bits` and `clear_executable_bits` function as well)

Reply:
thanks for the pointer, I realized we also need to add `Lib_root` and `Libexec_root`. I updated the code